File: support/infrastructure/src/main/java/no/officenet/example/rpm/support/infrastructure/jpa/LazyInitAspect.java
package no.officenet.example.rpm.support.infrastructure.jpa;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.Signature;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.reflect.FieldSignature;
import org.aspectj.lang.reflect.MethodSignature;
import org.hibernate.Hibernate;
import org.hibernate.HibernateException;
import org.hibernate.Session;
import org.hibernate.SessionFactory;
import org.hibernate.collection.AbstractPersistentCollection;
import org.hibernate.engine.SessionImplementor;
import org.hibernate.proxy.HibernateProxy;
import org.hibernate.proxy.LazyInitializer;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.orm.hibernate3.SessionFactoryUtils;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import java.io.Serializable;
import java.lang.reflect.Field;
import java.lang.reflect.Method;

@Aspect
public class LazyInitAspect implements InitializingBean {
	private static transient Log log = LogFactory.getLog(LazyInitAspect.class);

	private class ThreadLocaleInit<T> extends ThreadLocal<T> {
		protected T initVal;

		public ThreadLocaleInit(T initValue) {
			initVal = initValue;
		}

		@Override
		protected T initialValue() {
			return initVal;
		}
	}

	@PersistenceContext(unitName = "RPM")
	private EntityManager entityManagerRPM;
	@PersistenceContext(unitName = "PETS")
	private EntityManager entityManagerPETS;

	private SessionFactory sessionFactoryRPM;
	private SessionFactory sessionFactoryPETS;

	@Override
	public void afterPropertiesSet() throws Exception {
		sessionFactoryRPM = ((Session) entityManagerRPM.getDelegate()).getSessionFactory();
		sessionFactoryPETS = ((Session) entityManagerPETS.getDelegate()).getSessionFactory();
	}

	protected ThreadLocal<Boolean> locked = new ThreadLocaleInit<Boolean>(false);

	@Around("execution(* no.officenet.example.rpm.pets..*(..)) && no.officenet.example.rpm.support.infrastructure.spring.aop.SystemArchitectureAspect.lazyLoadableJpaProperties()")
	public Object reattachSessionPETS(ProceedingJoinPoint pjp) throws Throwable {
		return reattachSession(pjp, sessionFactoryPETS);
	}

	@Around("!execution(* no.officenet.example.rpm.pets..*(..)) && no.officenet.example.rpm.support.infrastructure.spring.aop.SystemArchitectureAspect.lazyLoadableJpaProperties()")
	public Object reattachSessionRPM(ProceedingJoinPoint pjp) throws Throwable {
		return reattachSession(pjp, sessionFactoryRPM);
	}

	public Object reattachSession(ProceedingJoinPoint pjp, SessionFactory sessionFactory) throws Throwable {
		Object obj = pjp.proceed();
		// Don't try to re-init if already re-initing some stuff on the same thread.
		// This happens when the persistence-provider calls the getters when it's initing)
		if (!locked.get()) {
			boolean reAttachNeeded = false;
			if (obj instanceof AbstractPersistentCollection) {
				AbstractPersistentCollection ps = (AbstractPersistentCollection) obj;
				reAttachNeeded = !ps.wasInitialized() && ps.getSession() == null;
			}

			if (obj instanceof HibernateProxy) {
				LazyInitializer li = ((HibernateProxy) obj).getHibernateLazyInitializer();
				reAttachNeeded = li.isUninitialized() && li.getSession() == null;
			}

			if (reAttachNeeded) {
				if (log.isTraceEnabled()) {
					log.trace("Re-attaching Hibernate session to " + pjp.toString());
				}
				Session session = null;
				locked.set(true);

				boolean newSession = false;
				try {
					try {
						session = sessionFactory.getCurrentSession();
					} catch (HibernateException e) {
						session = SessionFactoryUtils.getNewSession(sessionFactory);
						newSession = true;
					}

					SessionImplementor implementor = (SessionImplementor) session;

					Object pjpTarget = pjp.getTarget();
					Class entityType = pjpTarget.getClass();
					Serializable identifier = sessionFactory.getClassMetadata(entityType).getIdentifier(pjpTarget, implementor);

					Object newInstance = session.load(entityType, identifier);

					Signature signature = pjp.getSignature();
					if (signature instanceof MethodSignature) {
						MethodSignature methodSignature = (MethodSignature) signature;
						Method m = methodSignature.getMethod();
						boolean wasAccessible = m.isAccessible();
						if (!wasAccessible) {
							m.setAccessible(true);
						}
						obj = m.invoke(newInstance);
						if (!wasAccessible) {
							m.setAccessible(false);
						}
					} else if (signature instanceof FieldSignature) {
						// The instance will never be AbstractPersistentCollection because we've loaded an entity
						if (newInstance instanceof HibernateProxy) {
							newInstance = ((HibernateProxy)newInstance).getHibernateLazyInitializer().getImplementation();
						}
						FieldSignature fieldSignature = (FieldSignature) signature;
						Field field = fieldSignature.getField();
						boolean wasAccessible = field.isAccessible();
						if (!wasAccessible) {
							field.setAccessible(true);
						}
						obj = field.get(newInstance);
						// Stuff the updated field in the original object
						field.set(pjpTarget, obj);
						if (!wasAccessible) {
							field.setAccessible(false);
						}
					} else {
						throw new IllegalStateException("Unsupported signature-type: " + signature);
					}

					// Initialize the hole entity, else getters on primitives might throw exception after session-close
					Hibernate.initialize(obj);
				} finally {
					locked.set(false);
					if (session != null && newSession) {
						try {
							SessionFactoryUtils.closeSession(session);
						} catch (Throwable t) {
							// Fall thru
						}
					}
				}
			}
		}
		return obj;
	}
}

File: support/infrastructure/src/main/scala/no/officenet/example/rpm/support/infrastructure/spring/aop/LazyInitAspect.scala
package no.officenet.example.rpm.support.infrastructure.spring.aop

import org.apache.commons.logging.Log
import org.apache.commons.logging.LogFactory
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Pointcut
import org.aspectj.lang.reflect.FieldSignature
import org.aspectj.lang.reflect.MethodSignature
import org.hibernate.Hibernate
import org.hibernate.HibernateException
import org.hibernate.Session
import org.hibernate.SessionFactory
import org.hibernate.collection.AbstractPersistentCollection
import org.hibernate.engine.SessionImplementor
import org.hibernate.proxy.HibernateProxy
import org.springframework.beans.factory.InitializingBean
import org.springframework.orm.hibernate3.SessionFactoryUtils
import javax.persistence.EntityManager
import javax.persistence.PersistenceContext
import java.lang.reflect.Method

@Aspect
class LazyInitAspect extends InitializingBean {
	private val log: Log = LogFactory.getLog(getClass)

	@PersistenceContext
	private val entityManager: EntityManager = null
	private var sessionFactory: SessionFactory = null
	protected var locked = new ThreadLocalInit[Boolean](false)

	def afterPropertiesSet() {
		sessionFactory = (entityManager.getDelegate.asInstanceOf[Session]).getSessionFactory
	}

	@Pointcut("(@within(javax.persistence.Entity) || @within(javax.persistence.MappedSuperclass))" +
		" && (" +
		"@annotation(javax.persistence.ManyToOne)" +
		" || @annotation(javax.persistence.ManyToMany)" +
		" || @annotation(javax.persistence.OneToMany)" +
		" || @annotation(javax.persistence.OneToOne)" +
		")")
	def lazyLoadableJpaProperties() {
	}

	private def getPropertyName(method: Method): String = {
		val methodName = method.getName
		var propertyName: String = null
		if (methodName.startsWith("get")) {
			propertyName = methodName.substring("get".length)
		}
		else if (methodName.startsWith("is")) {
            // Note the getter will probably never be a "is" because this aspect only applies to
            // JPA-associations, which are not primitives, thus never boolean-types
			propertyName = methodName.substring("is".length)
		}
		else {
			throw new IllegalArgumentException("Unknow read-method: " + methodName)
		}
		propertyName = propertyName.substring(0, 1).toLowerCase + propertyName.substring(1)
		propertyName
	}

	private def getImplementation(obj: AnyRef): AnyRef = {
		Hibernate.initialize(obj)
		obj match {
			case hp: HibernateProxy => hp.getHibernateLazyInitializer.getImplementation
			case _ => obj
		}
	}

	private def getSetter(entityType: Class[_], propertyName: String, returnType: Class[_]): Method = {
		val setterName: String = "set" + propertyName.substring(0, 1).toUpperCase + propertyName.substring(1)
		entityType.getDeclaredMethod(setterName, returnType)
	}

	@Around("no.officenet.example.rpm.support.infrastructure.spring.aop.LazyInitAspect.lazyLoadableJpaProperties()")
	def reattachSession(pjp: ProceedingJoinPoint): AnyRef = {
		var obj: AnyRef = pjp.proceed
		if (obj == null || !LazyInitState.lazyInit.get) return obj
		val traceEnabled: Boolean = log.isTraceEnabled
		// Don't try to re-init if already re-initing some stuff on the same thread.
		// This happens when the persistence-provider calls the getters when it's initing)
		if (!locked.get) {
			val implementor: SessionImplementor = getSessionImplementor(obj)
			val hasOpenSession = implementor != null && implementor.isOpen
			val reAttachNeeded = !hasOpenSession && !Hibernate.isInitialized(obj)
			if (reAttachNeeded) {
				try {
					locked.set(true)
					obj = loadAndReAttach(pjp, traceEnabled)
				}
				finally {
					locked.set(false)
				}
			}
		}
		obj
	}

	private def getSessionImplementor(obj: AnyRef): SessionImplementor = {
		var implementor: SessionImplementor = null
		if (obj.isInstanceOf[AbstractPersistentCollection]) {
			val ps = obj.asInstanceOf[AbstractPersistentCollection]
			implementor = ps.getSession
		}
		else if (obj.isInstanceOf[HibernateProxy]) {
			val li = (obj.asInstanceOf[HibernateProxy]).getHibernateLazyInitializer
			implementor = li.getSession
		}
		implementor
	}

	private def loadAndReAttach(pjp: ProceedingJoinPoint, traceEnabled: Boolean): AnyRef = {
		var session: Session = null
		var obj: AnyRef = null
		var newSession = false
		try {
			try {
				session = sessionFactory.getCurrentSession
			}
			catch {
				case e: HibernateException => {
					session = SessionFactoryUtils.getNewSession(sessionFactory)
					newSession = true
				}
			}
			if (traceEnabled) {
				log.trace("Re-attaching Hibernate session to " + pjp.toString)
			}
			val implementor: SessionImplementor = session.asInstanceOf[SessionImplementor]
			val pjpTarget = pjp.getTarget
			val entityType = pjpTarget.getClass
			val identifier = sessionFactory.getClassMetadata(entityType).getIdentifier(pjpTarget, implementor)
			val newInstance = getImplementation(session.load(entityType, identifier))
			obj = initializeAndReAttach(pjp, newInstance)
		}
		finally {
			if (session != null && newSession) {
				try {
					SessionFactoryUtils.closeSession(session)
				}
				catch {
					case t: Throwable => {
					}
				}
			}
		}
		if (traceEnabled) {
			log.trace("obj: " + obj.getClass.getSimpleName + "@" + System.identityHashCode(obj) + " isInitialized: " + Hibernate.isInitialized(obj))
		}
		obj
	}

	private def initializeAndReAttach(pjp: ProceedingJoinPoint, newInstance: AnyRef): AnyRef = {
		var obj: AnyRef = null
		val pjpTarget = pjp.getTarget
		val entityType = pjpTarget.getClass
		pjp.getSignature match {
			case methodSignature: MethodSignature =>
				val getter = methodSignature.getMethod
				var wasAccessible = getter.isAccessible
				if (!wasAccessible) {
					getter.setAccessible(true)
				}
				obj = getter.invoke(newInstance)
				Hibernate.initialize(obj)
				if (!wasAccessible) {
					getter.setAccessible(false)
				}
				val propertyName: String = getPropertyName(getter)
				val setter: Method = getSetter(entityType, propertyName, getter.getReturnType)
				wasAccessible = setter.isAccessible
				if (!wasAccessible) {
					setter.setAccessible(true)
				}
				setter.invoke(pjpTarget, obj)
				if (!wasAccessible) {
					setter.setAccessible(false)
				}
			case fieldSignature: FieldSignature =>
				val field = fieldSignature.getField
				val wasAccessible = field.isAccessible
				if (!wasAccessible) {
					field.setAccessible(true)
				}
				obj = field.get(newInstance)
				Hibernate.initialize(obj)
				field.set(pjpTarget, obj)
				if (!wasAccessible) {
					field.setAccessible(false)
				}
				case signature =>
					throw new IllegalStateException("Unsupported signature-type: " + signature)
		}
		obj
	}

}